File: legacy/python/spc_up/services/export/origem.py
"""Build Origem de Recursos XML from confirmed entrada movimentacoes."""

from __future__ import annotations

from pathlib import Path

from lxml import etree
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from spc_up.config import settings
from spc_up.models.entities import (
    ContaBancaria,
    Movimentacao,
    MovimentacaoDirecao,
    MovimentacaoSpca,
    MovimentacaoStatus,
    PessoaFisica,
    PessoaJuridica,
)
from spc_up.services.export.common import (
    build_cabecalho,
    format_moeda,
    make_origem_root,
    sub,
    write_xml,
)

_ESPECIE_TAG = {
    "PIX": "transferenciaEletronicaPIX",
    "TED": "transferenciaEletronicaTED",
    "TEL": "transferenciaEletronicaTEL",
    "TEB": "transferenciaEletronicaTEB",
    "CH": "depositoCheque",
    "EP": "depositoEspecie",
    "OB": "ordemBancaria",
    "CC": "cartaoCredito",
    "AC": "avisoCredito",
    "OT": "outrosTitulosCredito",
}


def _fetch_movimentacoes(session: Session, uf: str, exercicio: int) -> list[Movimentacao]:
    return list(
        session.execute(
            select(Movimentacao)
            .join(MovimentacaoSpca, MovimentacaoSpca.movimentacao_id == Movimentacao.id)
            .options(
                selectinload(Movimentacao.spca),
                selectinload(Movimentacao.pessoa_fisica),
                selectinload(Movimentacao.pessoa_juridica),
                selectinload(Movimentacao.conta_bancaria),
            )
            .where(
                Movimentacao.uf == uf,
                Movimentacao.exercicio == exercicio,
                Movimentacao.direcao == MovimentacaoDirecao.ENTRADA.value,
                Movimentacao.status == MovimentacaoStatus.CONFIRMADO.value,
            )
            .order_by(Movimentacao.data_movimento, Movimentacao.id)
        ).scalars().all()
    )


def _append_conta_destino(parent: etree._Element, conta: ContaBancaria | None) -> None:
    conta_destino = sub(parent, "contaBancariaDestino")
    banco = sub(conta_destino, "bancoDestino")
    if conta is None:
        sub(banco, "nrBancoDestino", "000")
        sub(banco, "agenciaDestino", "0000")
        sub(banco, "contaCorrente", "0")
        sub(banco, "dvContaCorrente", "0")
        return

    sub(banco, "nrBancoDestino", "000")
    sub(banco, "agenciaDestino", conta.agencia)
    sub(banco, "contaCorrente", conta.conta)
    sub(banco, "dvContaCorrente", conta.dv or "0")


def _append_especie_recurso(
    parent: etree._Element,
    spca: MovimentacaoSpca,
    movimentacao: Movimentacao,
) -> None:
    especie = (spca.especie_recurso or "PIX").upper()
    tag = _ESPECIE_TAG.get(especie, "transferenciaEletronicaPIX")
    especie_wrapper = sub(parent, "especieRecurso")
    especie_node = sub(especie_wrapper, tag)
    sub(especie_node, "especieRecurso", especie)

    if movimentacao.nr_extrato_bancario:
        sub(especie_node, "nrExtratoBancario", movimentacao.nr_extrato_bancario)

    if tag in {
        "transferenciaEletronicaPIX",
        "transferenciaEletronicaTED",
        "transferenciaEletronicaTEL",
        "transferenciaEletronicaTEB",
        "depositoCheque",
        "avisoCredito",
        "depositoEspecie",
        "ordemBancaria",
        "cartaoCredito",
    }:
        _append_conta_destino(especie_node, movimentacao.conta_bancaria)


def _append_origem_recurso(
    parent: etree._Element,
    spca: MovimentacaoSpca,
    movimentacao: Movimentacao,
) -> None:
    origem_recurso = sub(parent, "origemRecurso")
    tipo = (spca.tipo_origem_recurso or "PF").upper()

    if tipo == "PF":
        pessoa: PessoaFisica | None = movimentacao.pessoa_fisica
        if pessoa is None:
            raise ValueError(f"Movimentacao {movimentacao.id} requires pessoa_fisica for PF origem")
        pf = sub(origem_recurso, "pessoaFisica")
        sub(pf, "tipo", "PF")
        sub(pf, "nrCpf", pessoa.cpf)
        sub(pf, "nmPessoa", pessoa.nome)
        if pessoa.titulo_eleitor:
            sub(pf, "tituloEleitor", pessoa.titulo_eleitor)
        return

    if tipo == "PJ":
        pessoa = movimentacao.pessoa_juridica
        if pessoa is None:
            raise ValueError(f"Movimentacao {movimentacao.id} requires pessoa_juridica for PJ origem")
        pj = sub(origem_recurso, "pessoaJuridica")
        sub(pj, "tipo", "PJ")
        sub(pj, "nrCnpj", pessoa.cnpj)
        sub(pj, "nmPessoa", pessoa.razao_social)
        return

    raise ValueError(f"Unsupported tipo_origem_recurso: {tipo}")


def _append_origem_item(parent: etree._Element, movimentacao: Movimentacao) -> None:
    spca = movimentacao.spca
    if spca is None:
        raise ValueError(f"Movimentacao {movimentacao.id} missing spca payload")

    origem = sub(parent, "origem")
    sub(origem, "dtEntrada", movimentacao.data_movimento.isoformat())
    sub(origem, "vrOrigem", format_moeda(movimentacao.valor))
    sub(origem, "fonteRecurso", spca.fonte_recurso)
    sub(origem, "naturezaRecurso", spca.natureza_recurso)
    _append_origem_recurso(origem, spca, movimentacao)
    sub(origem, "classificacaoReceita", spca.classificacao_receita)

    if spca.descricao_resumida:
        sub(origem, "descricaoResumida", spca.descricao_resumida)
    if spca.nr_recibo_doacao:
        sub(origem, "nrReciboDoacao", spca.nr_recibo_doacao)

    _append_especie_recurso(origem, spca, movimentacao)


def build_origem_xml(session: Session, uf: str, exercicio: int, cnpj: str) -> Path:
    """Build and persist Origem de Recursos XML for confirmed entradas."""
    movimentacoes = _fetch_movimentacoes(session, uf, exercicio)

    root = make_origem_root()
    build_cabecalho(root, cnpj=cnpj, exercicio=exercicio)

    corpo = sub(root, "CORPO")
    origens = sub(corpo, "origens")
    sub(origens, "totalOrigem", len(movimentacoes))

    for movimentacao in movimentacoes:
        _append_origem_item(origens, movimentacao)

    output_path = (
        Path(settings.storage_root)
        / "exports"
        / uf
        / str(exercicio)
        / f"origem_{cnpj}.xml"
    )
    return write_xml(root, output_path)

File: packages/core/src/export/origem.ts
import { and, asc, eq } from "drizzle-orm";

import {
  type ContaBancaria,
  type Db,
  type Movimentacao,
  type MovimentacaoSpca,
  movimentacao,
} from "@spc-up/db";

import {
  buildCabecalho,
  exportPath,
  formatMoeda,
  makeOrigemRoot,
  sub,
  writeXml,
  xmlToBuffer,
} from "./common";
import { scopePrestadorExercicio } from "./scope";

const ESPECIE_TAG: Record<string, string> = {
  PIX: "transferenciaEletronicaPIX",
  TED: "transferenciaEletronicaTED",
  TEL: "transferenciaEletronicaTEL",
  TEB: "transferenciaEletronicaTEB",
  CH: "depositoCheque",
  EP: "depositoEspecie",
  OB: "ordemBancaria",
  CC: "cartaoCredito",
  AC: "avisoCredito",
  OT: "outrosTitulosCredito",
};

const CONTA_DESTINO_TAGS = new Set([
  "transferenciaEletronicaPIX",
  "transferenciaEletronicaTED",
  "transferenciaEletronicaTEL",
  "transferenciaEletronicaTEB",
  "depositoCheque",
  "avisoCredito",
  "depositoEspecie",
  "ordemBancaria",
  "cartaoCredito",
]);

export type MovimentacaoOrigem = Movimentacao & {
  spca: MovimentacaoSpca;
  pessoaFisica: { cpf: string; nome: string; tituloEleitor: string | null } | null;
  pessoaJuridica: { cnpj: string; razaoSocial: string } | null;
  contaBancaria: ContaBancaria | null;
};

async function fetchMovimentacoes(
  db: Db,
  cnpjPrestador: string,
  exercicio: number,
): Promise<MovimentacaoOrigem[]> {
  const rows = await db.query.movimentacao.findMany({
    where: and(
      scopePrestadorExercicio(cnpjPrestador, exercicio),
      eq(movimentacao.direcao, "ENTRADA"),
      eq(movimentacao.status, "CONFIRMADO"),
    ),
    with: {
      spca: true,
      pessoaFisica: true,
      pessoaJuridica: true,
      contaBancaria: true,
    },
    orderBy: [asc(movimentacao.dataMovimento), asc(movimentacao.id)],
  });

  return rows.filter((row) => row.spca != null) as MovimentacaoOrigem[];
}

function formatDate(value: string | Date): string {
  if (typeof value === "string") {
    return value.slice(0, 10);
  }
  return value.toISOString().slice(0, 10);
}

function appendContaDestino(parent: ReturnType<typeof sub>, conta: ContaBancaria | null): void {
  const contaDestino = sub(parent, "contaBancariaDestino");
  const banco = sub(contaDestino, "bancoDestino");
  if (conta == null) {
    sub(banco, "nrBancoDestino", "000");
    sub(banco, "agenciaDestino", "0000");
    sub(banco, "contaCorrente", "0");
    sub(banco, "dvContaCorrente", "0");
    return;
  }
  sub(banco, "nrBancoDestino", "000");
  sub(banco, "agenciaDestino", conta.agencia);
  sub(banco, "contaCorrente", conta.conta);
  sub(banco, "dvContaCorrente", conta.dv ?? "0");
}

function appendEspecieRecurso(
  parent: ReturnType<typeof sub>,
  spca: MovimentacaoSpca,
  mov: MovimentacaoOrigem,
): void {
  const especie = (spca.especieRecurso ?? "PIX").toUpperCase();
  const tag = ESPECIE_TAG[especie] ?? "transferenciaEletronicaPIX";
  const especieWrapper = sub(parent, "especieRecurso");
  const especieNode = sub(especieWrapper, tag);
  sub(especieNode, "especieRecurso", especie);
  if (mov.nrExtratoBancario) {
    sub(especieNode, "nrExtratoBancario", mov.nrExtratoBancario);
  }
  if (CONTA_DESTINO_TAGS.has(tag)) {
    appendContaDestino(especieNode, mov.contaBancaria);
  }
}

function appendOrigemRecurso(
  parent: ReturnType<typeof sub>,
  spca: MovimentacaoSpca,
  mov: MovimentacaoOrigem,
): void {
  const origemRecurso = sub(parent, "origemRecurso");
  const tipo = (spca.tipoOrigemRecurso ?? "PF").toUpperCase();

  if (tipo === "PF") {
    const pessoa = mov.pessoaFisica;
    if (pessoa == null) {
      throw new Error(`Movimentacao ${mov.id} requires pessoa_fisica for PF origem`);
    }
    const pf = sub(origemRecurso, "pessoaFisica");
    sub(pf, "tipo", "PF");
    sub(pf, "nrCpf", pessoa.cpf);
    sub(pf, "nmPessoa", pessoa.nome);
    if (pessoa.tituloEleitor) {
      sub(pf, "tituloEleitor", pessoa.tituloEleitor);
    }
    return;
  }

  if (tipo === "PJ") {
    const pessoa = mov.pessoaJuridica;
    if (pessoa == null) {
      throw new Error(`Movimentacao ${mov.id} requires pessoa_juridica for PJ origem`);
    }
    const pj = sub(origemRecurso, "pessoaJuridica");
    sub(pj, "tipo", "PJ");
    sub(pj, "nrCnpj", pessoa.cnpj);
    sub(pj, "nmPessoa", pessoa.razaoSocial);
    return;
  }

  throw new Error(`Unsupported tipo_origem_recurso: ${tipo}`);
}

function appendOrigemItem(parent: ReturnType<typeof sub>, mov: MovimentacaoOrigem): void {
  const spca = mov.spca;
  const origem = sub(parent, "origem");
  sub(origem, "dtEntrada", formatDate(mov.dataMovimento));
  sub(origem, "vrOrigem", formatMoeda(mov.valor));
  sub(origem, "fonteRecurso", spca.fonteRecurso);
  sub(origem, "naturezaRecurso", spca.naturezaRecurso);
  appendOrigemRecurso(origem, spca, mov);
  sub(origem, "classificacaoReceita", spca.classificacaoReceita);
  if (spca.descricaoResumida) {
    sub(origem, "descricaoResumida", spca.descricaoResumida);
  }
  if (spca.nrReciboDoacao) {
    sub(origem, "nrReciboDoacao", spca.nrReciboDoacao);
  }
  appendEspecieRecurso(origem, spca, mov);
}

export function buildOrigemDocument(
  movimentacoes: MovimentacaoOrigem[],
  cnpj: string,
  exercicio: number,
) {
  const root = makeOrigemRoot();
  buildCabecalho(root, { cnpj, exercicio });
  const corpo = sub(root, "CORPO");
  const origens = sub(corpo, "origens");
  sub(origens, "totalOrigem", movimentacoes.length);
  for (const mov of movimentacoes) {
    appendOrigemItem(origens, mov);
  }
  return root;
}

/** Build and persist Origem de Recursos XML for confirmed entradas. */
export async function buildOrigemXml(
  db: Db,
  uf: string,
  exercicio: number,
  cnpj: string,
): Promise<string> {
  const movimentacoes = await fetchMovimentacoes(db, cnpj, exercicio);
  const doc = buildOrigemDocument(movimentacoes, cnpj, exercicio);
  return writeXml(doc, exportPath(uf.toUpperCase(), exercicio, cnpj, "origem"));
}

export async function buildOrigemXmlBuffer(
  db: Db,
  uf: string,
  exercicio: number,
  cnpj: string,
): Promise<Buffer> {
  const movimentacoes = await fetchMovimentacoes(db, cnpj, exercicio);
  return xmlToBuffer(buildOrigemDocument(movimentacoes, cnpj, exercicio));
}
